lightning-transaction-sync/src/electrum.rs

```diff
@@ -0,0 +1,448 @@
+				if tip_is_new {
+					// First check for any unconfirmed transactions and act on it immediately.
+					match self.get_unconfirmed_transactions(&confirmables) {
+						Ok(unconfirmed_txs) => {
+							// Double-check the tip hash. If it changed, a reorg happened since
+							// we started syncing and we need to restart last-minute.
+							if self.check_update_tip(&mut tip_header, &mut tip_height)?
+							{
+								log_debug!(self.logger,
+									"Encountered inconsistency during transaction sync, restarting.");
+								sync_state.pending_sync = true;
+								continue;
+							}
+
+							num_unconfirmed += unconfirmed_txs.len();
+							sync_state.sync_unconfirmed_transactions(&confirmables,
+								unconfirmed_txs);
+						},
+						Err(err) => {
+							// (Semi-)permanent failure, retry later.
+							log_error!(self.logger,
+								"Failed during transaction sync, aborting. Synced so far: {} confirmed, {} unconfirmed.",
+								num_confirmed,
+								num_unconfirmed
+							);
+							sync_state.pending_sync = true;
+							return Err(TxSyncError::from(err));
+						}
+					}
+
+					// Update the best block.
+					for c in &confirmables {
+						c.best_block_updated(&tip_header, tip_height);
+					}
+				}
+
+				match self.get_confirmed_transactions(&sync_state) {
+					Ok(confirmed_txs) => {
+						// Double-check the tip hash. If it changed, a reorg happened since
+						// we started syncing and we need to restart last-minute.
+						if self.check_update_tip(&mut tip_header, &mut tip_height)?
+						{
+							log_debug!(self.logger,
+								"Encountered inconsistency during transaction sync, restarting.");
+							sync_state.pending_sync = true;
+							continue;
+						}
+
+						num_confirmed += confirmed_txs.len();
+						sync_state.sync_confirmed_transactions(
+							&confirmables,
+							confirmed_txs,
+						);
+					}
+					Err(InternalError::Inconsistency) => {
+						// Immediately restart syncing when we encounter any inconsistencies.
+						log_debug!(self.logger,
+							"Encountered inconsistency during transaction sync, restarting.");
+						sync_state.pending_sync = true;
+						continue;
+					}
+					Err(err) => {
+						// (Semi-)permanent failure, retry later.
+						log_error!(self.logger,
+							"Failed during transaction sync, aborting. Synced so far: {} confirmed, {} unconfirmed.",
+							num_confirmed,
+							num_unconfirmed
+						);
+						sync_state.pending_sync = true;
+						return Err(TxSyncError::from(err));
+					}
+				}
+				sync_state.last_sync_hash = Some(tip_header.block_hash());
+				sync_state.pending_sync = false;
+			}
+		}
+		log_debug!(self.logger,
+			"Finished transaction sync at tip {} in {}ms: {} confirmed, {} unconfirmed.",
+			tip_header.block_hash(), start_time.elapsed().as_millis(), num_confirmed,
+			num_unconfirmed);
+		Ok(())
+	}
+
+	fn check_update_tip(&self, cur_tip_header: &mut BlockHeader, cur_tip_height: &mut u32)
```

Comment:
Maybe rename as `check_tip_updated` to make the result type clear.

Reply:
Ah, gotcha. Guess I was just reading it wrong lol. Could encode it in the return type similar to:

https://github.com/lightningdevkit/rust-lightning/blob/9c9e5f896c967d459d18d0c592c0a7a9efb3df98/lightning-block-sync/src/poll.rs#L20-L21

No need to change anything now as that is a larger change, but in general in-out parameters aren't as easy to reason about, IMO.